### ISSUE TITLE
EBU R128 Normalizes By Default to -26 dB

### ISSUE THREAD

**Issue by keithwoelke** [slhck/ffmpeg-normalize#48]
I'm not an expert, but shouldn't EBU R128 normalize to -23 dB by default?

**Comment by slhck** [comment]
Yes, you're right. I guess I'll change the default to -23 when EBU normalization is used.

-26 dB is just the default that was chosen for RMS-based normalization; it came from an ITU standardization activity where we first used this script.

PS: Even `ffmpeg` uses -24 as an Integrated Loudness target. Guess this should be changed as well?

**Comment by slhck** [comment]
Hm. The problem is: I would like to implement something like this:

```python
if ebu:
    if target == -26:
        target = -23
```

However, I can't detect whether the user has passed the option on the command line or whether the default was applied. This means that you could never set the integrated loudness target to -26 from the command line, because it would always be changed to -23.

I could of course set the normalization target to -23 for future versions altogether…?

**Comment by keithwoelke** [comment]
That's an interesting problem. I suppose you could set a flag before the default is applied, perhaps?

I suppose it wouldn't make sense to disallow the target for EBU normalization?

**Comment by slhck** [comment]
> I suppose you could set a flag before the default is applied, perhaps?

That's something that's handled by the `docopt` library that parses the options. But I could do some manual parsing like "if this flag is present in the option string" at the top of the script. That will probably work, although it's ugly.

> I suppose it wouldn't make sense to disallow the target for EBU normalization?

No, as people may want to deviate from the recommendation, I suppose. Note that the `loudnorm` filter is actually much more capable than what I've implemented in this wrapper script. I'd recommend people to actually use `ffmpeg` itself if they want to do EBU normalization, as that works in one pass.

**Comment by slhck** [comment]
For EBU normalization it's only one pass in ffmpeg. All other normalization
methods require two passes, which is why I wrote this script.

You don't need to be too familiar with ffmpeg in order to apply the
loudnorm filter, to be honest.

Am 09.01.2018 23:00 schrieb "Keith" <notifications@github.com>:

**Comment by michaelcrossland** [comment]
Just a heads up as of 01-05-18 ffmpeg still uses a 2 pass for EBU because
it still has to read the whole file before it can try to figure out how
much it needs to compress.

On Jan 9, 2018 4:00 PM, "Keith" <notifications@github.com> wrote:

The reason I am using this script is that I'm not very familiar with ffmpeg
nor normalization strategies. Does loudnorm use EBU by default? Does this
script do two passes?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/slhck/ffmpeg-normalize/issues/48#issuecomment-356428026>,
or mute the thread

**Comment by slhck** [comment]
Yes, to be precise: With loudnorm you need to call ffmpeg only once from
the command line.

Am 09.01.2018 23:04 schrieb "michaelcrossland" <notifications@github.com>:

> Just a heads up as of 01-05-18 ffmpeg still uses a 2 pass for EBU because
> it still has to read the whole file before it can try to figure out how
> much it needs to compress.
>

**Comment by michaelcrossland** [comment]
That is true. But as for ffmpeg it still does a 2 pass over the file one a
read the next a calculation pass when you do ebu from the command line.
It's just the nature of the Beast.

**Comment by michaelcrossland** [comment]
As for the other part of the script why don't you just have a 2nd function
just for when the ebu flag is used? Hell I can do that kind of thing in a
windows batch file. Yes I know I'm showing my age now. 😎

**Comment by michaelcrossland** [comment]
Oh would you mind if I worked with you here soon to wrap you script up in a
autoit program I'm making to make it ezer for windows users to use your
tool? Are you game I make the program and GUI and installer and windows
build of ffmpeg all wrapped around your script. What do you think?

**Comment by keithwoelke** [comment]
Part of the reason I felt overwhelmed trying to figure out loudnorm was because I referenced: http://k.ylo.ph/2016/04/04/loudnorm.html, which suggests two passes for files from the command-line, but that looked like a pain to script. Which is why I started looking for other solutions. I'm assuming your script isn't doing what is suggested here?

The command on the second pass looks like:

ffmpeg -i in.wav -af loudnorm=I=-16:TP=-1.5:LRA=11:measured_I=-27.61:measured_LRA=18.06:measured_TP=-4.47:measured_thresh=-39.20:offset=0.58:linear=true:print_format=summary

However, I couldn't figure out where a lot of those values even came from because it didn't seem like they were output by the first pass? And I have no idea where EBU fits into this equation. :-|

**Comment by michaelcrossland** [comment]
He is doing all of that. You just don't see it. He run that on the file
then python gets the output from ffmpeg then his script then does the math
to get you as close as ffmpeg can get to what ever level value you are
trying to get to.

On Jan 9, 2018 4:18 PM, "Keith" <notifications@github.com> wrote:

Part of the reason I felt overwhelmed trying to figure out loudnorm was
because I referenced: http://k.ylo.ph/2016/04/04/loudnorm.html, which
suggests two passes for files from the command-line, but that looked like a
pain to script. Which is why I started looking for other solutions. I'm
assuming your script isn't doing what is suggested here?

The command on the second pass looks like:

ffmpeg -i in.wav -af loudnorm=I=-16:TP=-1.5:LRA=11:
measured_I=-27.61:measured_LRA=18.06:measured_TP=-4.47:
measured_thresh=-39.20:offset=0.58:linear=true:print_format=summary

However, I couldn't figure out where a lot of those values even came from
because it didn't seem like they were output by the first pass? And I have
no idea where EBU fits into this equation. :-|

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/slhck/ffmpeg-normalize/issues/48#issuecomment-356432677>,
or mute the thread

**Comment by slhck** [comment]
Actually I'm not doing two manual passes with the `loudnorm` filter; I'm only doing that for peak and RMS normalization. Currently `loudnorm` in this script only does one pass. This is something I wanted to add for quite some time though. So @keithwoelke, you're right, it's a pain to script, but I guess I can implement it here. Just need a few hours on a weekend to do that.

**Comment by keithwoelke** [comment]
I got impatient. :-) Plus, I'm also compressing video files so I thought it might be better to do one pass to compress + normalize rather than separate passes for compression vs. normalization. Just in case it's remotely useful (probably not since you're using python):

target_i="-23"
target_tp="-1"
target_lra="15"

normalizationValues="$(ffmpeg -nostdin -i "$originalFile" -af loudnorm=I=${target_i}:TP=${target_tp}:LRA=${target_lra}:print_format=json -f null /dev/null 2>&1 | grep --null-data --perl-regexp --only-matching '{(.|\n)*}' | tr -d \\0)"

input_i="$(echo "$normalizationValues" | jq --raw-output ".input_i")"
input_tp="$(echo "$normalizationValues" | jq --raw-output ".input_tp")"
input_lra="$(echo "$normalizationValues" | jq --raw-output ".input_lra")"
input_thresh="$(echo "$normalizationValues" | jq --raw-output ".input_thresh")"
target_offset="$(echo "$normalizationValues" | jq --raw-output ".target_offset")"

ffmpeg -nostdin -y -i "$originalFile" -crf 23 -map_metadata -1 -metadata title="$title" -codec:s copy -s hd480 -af loudnorm=I=${target_i}:TP=${target_tp}:LRA=${target_lra}:measured_I=${input_i}:measured_LRA=${input_lra}:measured_TP=${input_tp}:measured_thresh=${input_thresh}:offset=${target_offset}:linear=true:print_format=summary -max_muxing_queue_size 1024 "$newFile"

**Comment by slhck** [comment]
Yes, for one pass that's absolutely enough 

 What I'm planning to do for the two pass case is an automatic parsing of the first pass results. I guess I have some time this weekend to prepare a 1.0 release of the program.

**Comment by keithwoelke** [comment]
Would your implementation give a different result? I assume you mean that you would do the first pass the same way you are now, take the summary, and reprocess the file again with the additional values instead of doing an analysis pass first?

**Comment by slhck** [comment]
For EBU normalization I would do actual two passes, so it would give a different result. The first pass would be done without video though.

For a one pass EBU normalization you basically  don't need ffmpeg-normalize.

**Comment by keithwoelke** [comment]
I'm wondering if maybe you didn't notice that my script does two passes?

The first is this line:

```
normalizationValues="$(ffmpeg -nostdin -i "$originalFile" -af loudnorm=I=${target_i}:TP=${target_tp}:LRA=${target_lra}:print_format=json -f null /dev/null 2>&1 | grep --null-data --perl-regexp --only-matching '{(.|\n)*}' | tr -d \\0)"
```

**Comment by slhck** [comment]
Oh, yes, you're right. I absolutely didn't see that properly on my mobile before. I can definitely use part of this code for what I'm trying to implement -- thanks!

**Comment by keithwoelke** [comment]
Glad to hear it. Here's the full file for my normalization script (stripped down version of the video script): 

https://github.com/keithwoelke/bash-scripts/blob/master/bin/normalize

I have cron jobs to kick off up to four ffmpeg processes hourly to committees and normalize

**Comment by slhck** [comment]
Version 1.0 now implements two-pass EBU R128 normalization and does that by default instead of peak or RMS normalization. (I've found it much more useful and subjectively more accurate in practice.)

Sorry for breaking existing scripts, but I had to change the command line options quite a bit in order to make the tool more flexible for possible future extensions.

Let me hear how it works for you!

**Comment by keithwoelke** [comment]
Awesome! Does it also skip normalizing if it's unneeded like the other methods do?

**Comment by slhck** [comment]
Not yet, I have to reintroduce a threshold feature for each individual audio stream. But I'm positive I'll get that done within the next few days.

**Comment by slhck** [comment]
Oh and I just realized it should be -23, not -24 as a default threshold.

**Comment by slhck** [comment]
Default is now -23 in v1.0.1. For threshold, the new issue is here: https://github.com/slhck/ffmpeg-normalize/issues/50

**Comment by sampumon** [comment]
Hi, sorry for slightly off-topicy resurrect here, but what is the difference _in result_ between:

1. single ffmpeg command two pass linear EBU,
2. ffmpeg-normalize actual two pass linear EBU?

@michaelcrossland wrote 25 days ago:
> Just a heads up as of 01-05-18 ffmpeg still uses a 2 pass for EBU because
it still has to read the whole file before it can try to figure out how
much it needs to compress.

@slhck wrote 15 days ago:
> For EBU normalization I would do actual two passes, so it would give a different result. The first pass would be done without video though.

Anyway, I `ffmpeg-normalize`'d 42 short films for screenings happening today, all went well so far : )

**Comment by slhck** [comment]
>  but what is the difference in result between:
>
> - single ffmpeg command two pass linear EBU,
> - ffmpeg-normalize actual two pass linear EBU?

You cannot do a “single ffmpeg command two pass linear EBU” normalization. FFmpeg always performs one pass only. My tool just helps in calling ffmpeg twice, and supplying the statistics of the first run as command line arguments to the second run.

For more info, please read [this blog article](http://k.ylo.ph/2016/04/04/loudnorm.html) from the person who wrote the `loudnorm` filter, especially:

> Supplying the filter with [the first pass] stats allow it to make more intelligent normalization decisions, as well as normalize linearly if possible. […] For precise integrated loudness, always use dual-pass mode and supply the offset parameter.

I'm glad it worked well for you – happy to see that people find the tool useful!

**Comment by creechy** [comment]
FWIW - I'm not sure the stated defaults (-nt ebu, -t -23) are actually the defaults. I ran the following two commands using version 1.2.1

ffmpeg-normalize -nt ebu hq0.mp4 -c:a aac -f -o hq0-norm-def.mp4
ffmpeg-normalize -nt ebu hq0.mp4 -c:a aac -nt ebu -t -23 -f -o hq0-norm-ebu-23.mp4

Then using open-source loudness-scanner I get different results

% loudness scan hq0-norm-def.mp4 hq0-norm-ebu-23.mp4 
  Loudness                                                                      
{-24.0} LUFS, hq0-norm-def.mp4
{-23.0} LUFS, hq0-norm-ebu-23.mp4

Thoughts?

**Comment by slhck** [comment]
See, I wrote that the default would be -23 LUFS, but I copied it over from ffmpeg's `loudnorm`, which uses -24. Since the actual [R128 standard](https://tech.ebu.ch/docs/r/r128.pdf) uses -23, I actually wanted to use -23 all along.

Thanks for catching that bug; fixed in 1.2.2!

**Comment by michaelcrossland** [comment]
That because the EBU does a full normalization so if there are really low
lows and really high highs it takes those drops and pikes into account so
as not to cuss clipping. So the out come will be with in + or - 5dB. But
over all it will "sound" to be around -26 to the ear. Only case I've found
where that is a draw back is when you her movie's that have 2.0 sound that
are taken from 5.1 or 7.1 sound in those cases it will make the sound over
all lower because in most cases the LEF ie subwoofer Channel get warped
into the 2.0 sound and the channel is amped up so that your subwoofer will
play the bass at about the right levels when played with the other channels.
